source/adios2/engine/insitumpi/InSituMPISchedules.cpp

```diff
@@ -0,0 +1,318 @@
+        for (auto &subFileIndexPair : variableNamePair.second)
+        {
+            // <steps, <SubFileInfo>>  but there is only one step
+            for (auto &stepPair : subFileIndexPair.second)
+            {
+                // <SubFileInfo>
+                for (auto &sfi : stepPair.second)
+                {
+                    n++;
+                }
+            }
+        }
+    }
+    return n;
+}
+
+int FixSeeksToZeroOffset(
+    std::map<std::string, SubFileInfoMap> &variablesSubFileInfo) noexcept
+{
+    int n;
+    for (auto &variableNamePair : variablesSubFileInfo)
+    {
+        // <writer, <steps, <SubFileInfo>>>
+        for (auto &subFileIndexPair : variableNamePair.second)
+        {
+            // <steps, <SubFileInfo>>  but there is only one step
+            for (auto &stepPair : subFileIndexPair.second)
+            {
+                // <SubFileInfo>
+                for (auto &sfi : stepPair.second)
+                {
+                    FixSeeksToZeroOffset(sfi);
+                    n++;
+                }
+            }
+        }
+    }
+    return n;
+}
+
+void FixSeeksToZeroOffset(SubFileInfo &record) noexcept
+{
+    size_t nElements = record.Seeks.second - record.Seeks.first + 1;
+    size_t pos =
+        LinearIndex(record.BlockBox, record.IntersectionBox.first, true);
+    record.Seeks.first = pos;
+    record.Seeks.second = pos + nElements - 1;
+}
+
+std::vector<std::vector<char>> SerializeLocalReadSchedule(
+    const int nWriters,
+    const std::map<std::string, SubFileInfoMap> &variablesSubFileInfo) noexcept
+{
+    std::vector<std::vector<char>> buffers(nWriters);
+
+    // Create a buffer for each writer
+    std::vector<int> nVarPerWriter(nWriters);
+    for (size_t i = 0; i < nWriters; i++)
+    {
+        nVarPerWriter[i] = 0;
+        InsertToBuffer(buffers[i], &nVarPerWriter[i],
+                       1); // allocate first 4 bytes
+    }
+
+    for (const auto &variableNamePair : variablesSubFileInfo)
+    {
+        const std::string variableName(variableNamePair.first);
+        // <writer, <steps, <SubFileInfo>>>
+        for (const auto &subFileIndexPair : variableNamePair.second)
+        {
+            const size_t subFileIndex = subFileIndexPair.first; // writer
+            auto &lrs = buffers[subFileIndex];
+            // <steps, <SubFileInfo>>  but there is only one step
+            for (const auto &stepPair : subFileIndexPair.second)
+            {
+                // LocalReadSchedule sfi = subFileIndexPair.second[0];
+                const std::vector<SubFileInfo> &sfi = stepPair.second;
+                SerializeLocalReadSchedule(buffers[subFileIndex], variableName,
+                                           sfi);
+                ++nVarPerWriter[subFileIndex];
+                break; // there is only one step here
+            }
+        }
+    }
+
+    // Record the number of actually requested variables for each buffer
+    for (size_t i = 0; i < nWriters; i++)
```

Comment:
I hate the multiple compiler warnings when index type and size type don't match. Happens to me all the time......stick with int or auto for i, or do static_cast<size_t>(nWriters).....either way works.